### PR TITLE
Add Dimensions controls to Product Collection block

### LINKED ISSUE #47884: Add "Width" and "Height" settings to Woo blocks nested inside Group/Row/Stack/Grid
We don't support the Width, Height, Min. Width, and Min. Height controls like most core blocks when nested inside a container block (Group, Row, Stack, Grid). Let's start by adding them to the Product Collection block. The new controls should be displayed in the Dimensions section in the Inspector panel.

image

### PR DESCRIPTION
**Pull request by octaedro**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR introduces `Width` controls to the `Product Collection` block, enabling users to define fixed or flexible width options.

The implementation adds the width configuration in the block inspector. The controls are placed at the `Product Collection` level and affect the block and its children. So width configuration will affect the `Product Template`, the `Pagination` and the `No Results` blocks.

**Why this change is needed**: This feature enhances flexibility in product display customization, allowing users to better manage the presentation of their product. 

https://github.com/user-attachments/assets/b9678179-49ed-45a1-b00e-3807baf07b28

Closes #47884.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a few products.
2. Navigate to Pages > Add New Page and insert a `Product Collection` block.
3. In the block setup dialog, select `create your own` option.
4. Press the settings button to open the inspector and scroll down to the `Dimensions` options.
5. Under the `Width` options, select `Fixed` and add a value (e.g. 400px).
6. Verify that the `Product Template` block reflects this width immediately.
7. Modify the fixed value and verify that the width changes immediately again.
8. Select the `Fill` option. The `Product Collection` block and its children should fill the available width.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx [hunk 1]
```diff
@@ -16,7 +16,6 @@ import {
 } from '@wordpress/block-editor';
 import { Spinner } from '@wordpress/components';
 import { store as coreStore } from '@wordpress/core-data';
-import { ProductCollectionAttributes } from '@woocommerce/blocks/product-collection/types';
```
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx]: I think it's in use so should be brought back.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php [hunk 2]
```diff
@@ -136,6 +136,45 @@ function ( $html ) {
...
+	 * Get the styles for the list element (fixed width).
+	 *
+	 * @param string $fixed_width Fixed width value.
+	 * @return string
+	 */
+	protected function get_list_styles( $fixed_width ) {
+		$style = '';
+
+		if ( isset( $fixed_width ) ) {
+			$style .= sprintf( 'width:%dpx;', intval( $fixed_width ) );
```
Comment (kmanijak) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php]: Control allows you to change units so we should not assume it's `px`. BTW unit is respected in Editor

image

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/width-options-control.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+				...dimensions,
+				widthType: type,
+			},
+		} );
+	};
+
+	return (
+		<ToolsPanelItem
+			label={ __( 'Width', 'woocommerce' ) }
+			hasValue={ () => true }
```
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/width-options-control.tsx]: I suggest using actual conditions for `hasValue` so basically whenever the value is different from a default:

```suggestion
			hasValue={ () => widthType !== WidthOptions.FILL }
			isShownByDefault
```

You can even import the default value from `plugins/woocommerce-blocks/assets/js/blocks/product-collection/constants.ts` for comparison.

If you wanted to make sure option is shown by default, use `isShownByDefault` prop as in the suggestion above.

**Review by kmanijak — CHANGES_REQUESTED** [review]
Hey, thanks for taking care of it! I tested the changes and found some issues but we're getting there!

One more thing is the Product Collection is left-aligned in Editor and centered in frontend when fixed width is applied

Editor | Frontend
-- | --
image | Screenshot 2024-09-26 at 12 11 24

I don't have preferences in which way but they should be aligned 🤔 I guess centered makes more sense and brings more value until we add some alignment options.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php [hunk 4]
```diff
@@ -392,8 +434,7 @@ public function enhance_product_collection_with_interactivity( $block_content, $
 				$block_content = $this->enable_client_side_navigation( $block_content );
 			}
 		}
-
-		return $block_content;
+		return sprintf( '<div>%s</div>', $block_content );
```
Comment (octaedro) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:437]: We add this `div` to center the product list on the front end when we select a percentage (%) `width`.

File: plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx (hunk 1 above)
Comment (octaedro) [reply plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx]: Good eyes! Added in the commit [cb8407d](https://github.com/woocommerce/woocommerce/pull/51420/commits/cb8407d66ce45e22bfa6da39ee36cd086f423f48)

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php (hunk 2 above)
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php]: Good call! I fixed this in the commit [1620002](https://github.com/woocommerce/woocommerce/pull/51420/commits/1620002ff43b43b40ef7e02ede420333ed8f7b52)

**Comment by octaedro** [comment]
Thank you @kmanijak for your review,

> I don't have preferences in which way but they should be aligned 🤔 I guess centered makes more sense and brings more value until we add some alignment options.

Good idea, I centered the items.

I addressed the changes you suggested and fixed the E2E tests. Could you take another look at this PR?

**Review by kmanijak — APPROVED** [review]
Thanks for addressing all of the comments, LGTM! Amazing work @octaedro! 🚀

**Pull request merged**